### PR TITLE
Pylance settings page

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
- Pylance settings page.
- Text Editor | Advanced page.
- Plumb setting changes to LSP.
- Don't fire 'changed' event when settings didn't change.
- Allow workspace override for selected options.

image

image

### CODE REVIEW

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 1]
```diff
@@ -158,24 +163,22 @@ public Task<TResult> InvokeWithParameterObjectAsync<TResult>(string targetName,
...
             var settings = new Settings {
                 python = new Settings.PythonSettings {
                     pythonPath = _clientContext.InterpreterConfiguration.InterpreterPath,
                     venvPath = string.Empty,
                     analysis = new Settings.PythonSettings.PythonAnalysisSettings {
-                        logLevel = "log",
-                        autoSearchPaths = false,
-                        diagnosticMode = "openFilesOnly",
-                        diagnosticSeverityOverrides = new Dictionary<string, string>(),
-                        extraPaths = _clientContext.SearchPaths.ToArray(),
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:176]: We shouldn't remove support for `extraPaths`. What was there for it looked correct.

**Comment by huguesv** [comment]
Rename the tree view item on the left from "Diagnostics" to "Analysis" ?

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs (hunk 1 above)
Comment (MikhailArkhipov) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:176]: Yes, I need to look into few more (and probably expose them), settings are scattered between Pylance, Pyright and core extension.

Comment (MikhailArkhipov) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:176]: Added more options

File: Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                changed = true;
+            }
+
+            var typeCheckingMode = _service.LoadString(nameof(TypeCheckingMode), Category) ?? PythonLanguageClient.TypeCheckingMode.Basic;
+            if (TypeCheckingMode != typeCheckingMode) {
+                TypeCheckingMode = typeCheckingMode;
+                changed = true;
+            }
+
+            var typeshedPaths = _service.LoadString(nameof(TypeshedPaths), Category)?.Split(';');
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs:73]: If it's a multi-line text box, users may be tempted to put newlines in there. Should we split on `';\r\n'` for extra convenience?

Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs:73]: oh wrong place

Comment (MikhailArkhipov) [reply Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs:73]: Yes, makes sense. I probably also need to add some text to the page wrt 'use line breaks or semicolon to separate paths'. Long term I am going to open task to implement better UI like

image

**Review by huguesv — APPROVED** [review]
If you haven't already, double check that the tab order is correct in both pages, since that is usually tricky when using the table control.

If we don't have a task for watching for workspace settings changed, we should create one.

File: Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply Python/Product/PythonTools/PythonTools/Options/PythonAnalysisOptions.cs]: Done.

**Pull request merged**